### ISSUE TITLE
proposal: add ProviderClient.AuthResult() method

### ISSUE THREAD

**Issue by majewsky** [gophercloud/gophercloud#1141]
ProviderClient (and hence, ServiceClient) is responsible for obtaining a Keystone token (i.e. sending the `POST /v3/auth/tokens` request to Keystone), but it only allows very limited access to the response to that request. The token itself can be retrieved via `ProviderClient.Token()`, and the service catalog is indirectly accessible via `ProviderClient.EndpointLocator`, but that's it.

There are other usecases where you need to look at the token:

1. The response contains the list of roles that you have, which may be interesting to selectively enable/disable functionality in the client.

2. My own usecase (which prompted this proposal) is that I'm building a service that needs to manipulate role assignments for its own service user. It therefore needs to know its own user ID. It gets its configuration from the OS_* env vars (via `clientconfig.AuthenticatedClient()`), so it could look at those vars, but that's not guaranteed to work.

    - Option 1: `OS_USER_ID` is supplied, and that's what we need.
    - Option 2: `OS_USERNAME` is supplied, but the service user may lack permission to list users, so it cannot resolve its own user name into an ID.
    - Option 3: `OS_TOKEN` is supplied. No way to transform this into a user directly.

    However, the auth response already directly contains the user ID in question, so no guessing is needed.

### Proposal

**NOTE: Cannot be implemented like this, [see below for amendments](https://github.com/gophercloud/gophercloud/issues/1141#issuecomment-415388044).**

Add the following API in the toplevel package:

```
func (p *ProviderClient) AuthResult() AuthResult { ... }

type AuthResult struct {
  V2Result *Result //response from POST /v2/tokens
  V3Result *Result //response from POST /v3/auth/tokens
}
```

The structure of the AuthResult type encodes the information which Identity API version was used to authenticate: Either all fields are nil (failed authentication), or exactly one field is non-nil (depending on which Identity API version was used). It is therefore not possible to e.g. inadvertently decode a V2 response with a V3 result type. The type is designed to allow simple future expansion when a new Identity API version is released.

Add the following API in package `openstack/identity/v3/tokens` (and analogously for v2):

```
func FromAuthResult(p *ProviderClient) *CreateResult {
  result := p.AuthResult().V3Result
  if result == nil {
    return nil
  }
  return &CreateResult{ commonResult { *result } }
}
```

This indirection is needed because `type AuthResult` cannot directly reference `tokens.CreateResult` as the type of its `V3Result` field. (That would create an import cycle.)

Example application for my usecase described above:

```
provider, err := clientconfig.AuthenticatedClient(nil)
//... ensure err != nil ...
result := tokens.FromAuthResult(provider)
//... ensure result != nil ...
token, err := result.ExtractToken()
//... ensure err != nil ...
log.Println("my own user ID is", token.User.ID)
```

**Comment by majewsky** [comment]
@jtopjian Can you have a look, please?

**Comment by jtopjian** [comment]
@majewsky I apologize - I have no idea how I've missed this.

This sounds like an appropriate thing to implement. Since this information is already being sent back during authentication, then I see no reason why we shouldn't make it accessible.

At some point I'm going to have to look at how the ProviderClient handles the service catalog and possibly do some automatic version discovery. I wonder if that version discovery and your proposed addition are related.

**Comment by majewsky** [comment]
What kind of automatic version discovery are you referring to? (As in: Whose version needs to be discovered?)

**Comment by jtopjian** [comment]
Parsing the service catalog for microversions - min and max supported, and then setting the client appropriately. It's not a high priority but if the Python clients begin implementing some of the stuff being designed in the various OpenStack API specs, it'll be something I need to look at down the road.

**Comment by majewsky** [comment]
I see, that makes sense. Do you have any comments on the proposed API? Otherwise I'd like to start implementing it.

**Comment by jtopjian** [comment]
> Do you have any comments on the proposed API?

I think it's a great idea overall and I think the core design will work.

Rather than have the functions in the `tokens` packages, what about having it in the `openstack` package? I think this functionality would fit well in the `openstack/clients.go` file if that's possible. Maybe by leveraging the same logic done here: https://github.com/gophercloud/gophercloud/blob/master/openstack/client.go#L89

Also, we might want to change the names. For example:

* `AuthResult()`: maybe `GetAuthScope()` or `GetScopedToken()` or `GetSomethingSomething()`
* `FromAuthResult()`: maybe use the same name as the Provider function?

I don't want to spend too much time on the right name, but I'd like to see the name somewhat align with other patterns used in Gophercloud.

Thoughts?

**Comment by majewsky** [comment]
> Rather than have the functions in the `tokens` packages, what about having it in the `openstack` package?

I don't have a strong opinion. As I said, it cannot go in package gophercloud because of import cycles, but package openstack would work afaics.

> Also, we might want to change the names.

> `AuthResult()`: maybe `GetAuthScope()` or `GetScopedToken()` or `GetSomethingSomething()`

Would `GetAuthResult` be okay for you? `AuthScope` already exists and is something very different. And it's not a token either, it's a `Result` containing a token.

> `FromAuthResult()`: maybe use the same name as the Provider function?

So... like this?

```
func GetAuthResult(provider *gophercloud.ProviderClient) gophercloud.Result {
    //return value is either tokens2.CreateResult or tokens3.CreateResult
    ...
}
```

**Comment by majewsky** [comment]
Tell you what, new proposal. I'll just put everything in one snippet this time because, as you might be able to tell from my writing style, I'm pretty exhausted at this point. 😐 

```
// ---------- package gophercloud ----------

type AuthResult interface {
    ExtractTokenID() (string, error)
    ExtractEndpointLocator() (EndpointLocator, error)
}

func (p *ProviderClient) AuthResult() AuthResult { return p.authResult }

//Deprecate SetToken() and direct write access to TokenID and EndpointLocator in favor of this new method:
func (p *.ProviderClient) SetAuthResult(r AuthResult) error {
    // ... lock reauth mutex like SetToken() does ...
    p.authResult = r
    p.TokenID, err = r.ExtractTokenID()
    p.EndpointLocator, err = r.ExtractEndpointLocator()
    // ... return err or nil ...
}

// ---------- package openstack ----------

type AuthResult struct {
```

**Comment by jtopjian** [comment]
> Edit: I just looked at type ProviderClient. None of the accessor methods have a Get prefix. So it would actually be more consistent to go with just AuthResult().

Yeah, that's fine. I was more focused on the name being a verb, not necessarily the word "Get" specifically. But you're right - some of the method/function names in the provider client are also nouns.

> Edit 2: Actually, that won't work either. When converting from *gophercloud.Result into *tokens.createResult, I need access to the package-private type tokens.commonResult, so the function doing that must be in package tokens.

We can change `commonResult` to `CommonResult`. 

> Edit 3: Okay, it's even worse. Inside the reauthfunc (i.e. in v2auth and v3auth)

I apologize - I was unclear with my suggestion. It wasn't to re-use the [`Authenticate`](https://github.com/gophercloud/gophercloud/blob/master/openstack/client.go#L89) function, but to use the [same logic within the function](https://github.com/gophercloud/gophercloud/blob/master/openstack/client.go#L95-L108). My thinking was that instead of using the `v2/tokens` and `v3/tokens` package, you could have one function in the `openstack` package which uses a switch statement based on the detected Identity endpoint.

But if integrating it in `Authenticate` would also work and would be beneficial, we can look at doing that, too.

> I have the feeling like I'm going through the same thought processes that I went through last month when writing the proposal

Totally understandable as you've spent much longer investigating this than I have 😄 

Let me know if the above comments help at all. Either way, I'll defer to your judgement on how you best see this implemented. If you want to put together a full working example, we can look at it from there.

**Comment by majewsky** [comment]
> I was more focused on the name being a verb, not necessarily the word "Get" specifically.

Ah, you mean that `AuthResult` could be misread as an action "authenticate result"?

> We can change commonResult to CommonResult.

Ok, then this would probably be the preferred route.

> [one function in package openstack with switch statement]

I see. That would be fine by me.

> But if integrating it in Authenticate would also work and would be beneficial, we can look at doing that, too.

I mean, it's architecturally very nice. It cleans up the boundary between the ProviderClient type and the auth code in `package openstack` in a way that would allow to make more things in ProviderClient private when the next backwards-incompatible release happens.

However, the original proposal (amended to make `tokens3.CommonResult` public) would be much more straight-forward. And I guess we both have better things to do right now. :)

**Comment by jtopjian** [comment]
Implemented in #1226
